### PR TITLE
We can now use a custom DatePicker component

### PR DESCRIPTION
**Pull request by SudoPlz**
Need this to be able to replace DatePickerIOS w/ something like dgladkov/react-native-uncontrolled-date-picker-ios

Usage:

```javascript
<TimerPicker
	pickerRefCb={(datePicker) => { this.datepickerRef = datePicker; }}
	customDatePickerIOS={Platform.OS === 'ios' ? UncontrolledDatePickerIOS : null}
	onConfirm={() => {
		if (this.datepickerRef && this.datepickerRef.getDate) {
			this.datepickerRef.getDate((datePicked) => {
				// do something with the datePicked Date obj
			});
		}
	}}
/>
```

### CODE REVIEW

**Comment by SudoPlz** [comment]
This allows us to use a custom date picker like dgladkov/react-native-uncontrolled-date-picker-ios.

Why?  Because we need it to get over this: https://github.com/facebook/react-native/issues/8169

**Comment by mmazzarolo** [comment]
Thanks @SudoPlz !
I'll wait a bit before merging this PR since the configuration of this component is getting a bit too crowded.  
Let's see if anyone else is interested in this feature :)

**Comment by abarisic86** [comment]
@SudoPlz is this tested with https://github.com/dgladkov/react-native-uncontrolled-date-picker-ios latest version? Because `uncontrolled-date-picker-ios` does not do anything with `onDateChange` method. Rather, it implements getDate() method that should be called inside `onConfirm` handler of the parent component.

**Comment by SudoPlz** [comment]
@abarisic86 I'm using this PR w/ https://github.com/dgladkov/react-native-uncontrolled-date-picker-ios on production already.

Notice the `pickerRefCb` prop? I use that like so:

```
<TimerPicker
    pickerRefCb={(datePicker) => { this.datepickerRef = datePicker; }}
    ...
```

and then `onConfirm` I do:
```
this.datepickerRef.getDate((jsDateTimeSelected) => {
    // do something with the date
});
```

**Comment by abarisic86** [comment]
@SudoPlz  it only works if I set `neverDisableConfirmIOS` to `true`. If I don't, it never calls `onConfirm` function.

Can you explain this line?
`<View onStartShouldSetResponderCapture={neverDisableConfirmIOS !== true ? this._handleUserTouchInit : null}>`

**Comment by SudoPlz** [comment]
Before my pull request, we used to disable the confirm button while the user was interacting with the picker.
`onStartShouldSetResponderCapture={this._handleUserTouchInit}` does exactly that.

Now because we'll be using `https://github.com/dgladkov/react-native-uncontrolled-date-picker-ios` we no longer need the confirm and cancel buttons to become disabled, thus we pass the `neverDisableConfirmIOS` flag as true.

Some people may still want that behaviour so it's optional.

`neverDisableConfirmIOS` doesn't look like it has the power to stop the `onConfirm` method from getting called though.

**Comment by abarisic86** [comment]
Yes, the usage of this parameter is clear. But is setting `neverDisableConfirmIOS = {true}` necessary for custom date picker to work? As I mentioned before, _handleDateChange inside modal is never called thus if 'neverDisableConfirmIOS' is set to true `userIsInteractingWithPicker` is never set to false and you can't click on confirm button.

The proper solution would be to modify `react-native-uncontrolled-date-picker-ios` to call `this._onDateChange` function when native picker signals that date is changed. A provided date will be wrong sometimes (that is the bug we are trying to solve) but we need that event fired to reset `userIsInteractingWithPicker` prop.

**Comment by SudoPlz** [comment]
Or quit using the `userIsInteractingWithPicker` functionality all together (which is a feature I added a long time ago to counter the exact problem that `https://github.com/dgladkov/react-native-uncontrolled-date-picker-ios` is solving now).

We don't need `userIsInteractingWithPicker` any more.
Also, yes, it's true, if you want to use a `customDatePickerIOS` you'll have to set `neverDisableConfirmIOS` to true as well. I prefer using that flag, from getting **any** wrong results at all.

Any reason why you don't want to use that flag?

**Comment by abarisic86** [comment]
Yes, I had to add `neverDisableConfirmIOS` and it will go to production in our app.

Option to disable confirm button while interacting with date picker is still useful to stop users that click on confirm before the picker is done spinning.

1. User opens the modal
2. Scroll few times, datePicker settles on some date (change recorded)
3. Makes another scroll gesture and clicks on confirm before date picker settles 
4. Date will change to value selected at step 2

I admit it is just an edge case and it's not the way datePicker should be used. But, if we want to fix this component for good - this option should be covered.

If we leave it as is I would **suggest**:
- there should be a note inside Readme.md with information that this won't work if `neverDisableConfirmIOS` isn't set to `true`
- add an explanation how to use customDatePickerIOS (something similar to your first post here)
- maybe move the code from line 155 to `_handleUserTouchInit` and use something like code below to indicate that if you are using custom datePicker `userIsInteractingWithPicker` should not change. In that case, this will work even if `neverDisableConfirmIOS` is set to false

`
 _handleUserTouchInit = () => {
    // custom date picker shouldn't change this param
    if(!customDatePickerIOS){
        this.setState({
            userIsInteractingWithPicker: true,
        })
    }
    return false;
};
`

**Comment by SudoPlz** [comment]
@abarisic86 That's a good suggestion, I'll do that.

Also, are you sure your steps are correct?
I haven't tested this, but to me it sounds as if it should be that way:

1. User opens the modal
2. Scroll few times, datePicker settles on some date (change recorded)
3. Makes another scroll gesture and clicks on confirm before date picker settles
4. ~Date will change to value selected at step 2~ `.getDate` will run, and whatever the current value of the spinning component is will be returned.

**Comment by SudoPlz** [comment]
All set. @mmazzarolo Can you merge? This doesn't change anything to the existing usage, it only adds an extra feature that lot's of people require.

File: src/CustomDatePickerIOS/index.js [hunk 1]
```diff
@@ -148,8 +162,9 @@ export default class CustomDatePickerIOS extends PureComponent {
       >
         <View style={[styles.datepickerContainer, datePickerContainerStyleIOS]}>
           {customTitleContainerIOS || titleContainer}
-          <View onStartShouldSetResponderCapture={this._handleUserTouchInit}>
-            <DatePickerIOS
+          <View onStartShouldSetResponderCapture={neverDisableConfirmIOS !== true ? this._handleUserTouchInit : null}>
```
Comment (abarisic86) [inline src/CustomDatePickerIOS/index.js]: You should revert this to ` <View onStartShouldSetResponderCapture={this._handleUserTouchInit}>`

**Comment by mmazzarolo** [comment]
Hey @SudoPlz and @abarisic86 
Even if I din't post here I was following the issue, thanks for the discussion!

@SudoPlz could you address the last @abarisic86 suggestion before the merge?

File: src/CustomDatePickerIOS/index.js (hunk 1 above)
Comment (SudoPlz) [reply src/CustomDatePickerIOS/index.js]: Hmm any reason why we should do that?
People may want to disable the confirmation buttons even if they're not using a Custom date picker.

Can you think of any reasons to revert?

Comment (abarisic86) [reply src/CustomDatePickerIOS/index.js]: Is there any reason to make this change? any scenario where we need this condition?

Although it doesn't generate any problems:
- neverDisableConfirmIOS feature works without this condition (when using default datepicker)
- The main problem I see is readability. It's easier to follow code if we leave this condition out
- "neverDisableConfirmIOS !== true" is confusing

So, can we leave it out?

**Comment by konstantin-mohin** [comment]
@mmazzarolo 
in the `_handleUserTouchInit` method use `this.pros.customDatePickerIOS`

**Comment by SudoPlz** [comment]
Ahm, but I did make those changes back then:

https://github.com/SudoPlz/react-native-modal-datetime-picker/blob/2652664de92c3b6968d4cbd3d836b7347159b72c/src/CustomDatePickerIOS/index.js#L88

**Pull request merged**